services/agents/channel/channel.go

```diff
@@ -280,23 +299,46 @@ func (c *Channel) subscribe(id uint32) chan agentpb.AgentResponsePayload {
 	return ch
 }
 
-func (c *Channel) publish(id uint32, resp agentpb.AgentResponsePayload) {
+func (c *Channel) removeResponseChannel(id uint32) (chan Response, error) {
 	c.rw.Lock()
-	if c.responses == nil { // Channel is closed, no more publishing
+	if c.responses == nil { // Channel is closed
 		c.rw.Unlock()
-		return
+		return nil, nil
 	}
 
 	ch := c.responses[id]
+	c.rw.Unlock()
 	if ch == nil {
-		c.rw.Unlock()
-		c.close(errors.WithStack(fmt.Errorf("no subscriber for ID %d", id)))
-		return
+		return nil, errors.WithStack(fmt.Errorf("no subscriber for ID %d", id))
 	}
-
+	c.rw.Lock()
 	delete(c.responses, id)
 	c.rw.Unlock()
-	ch <- resp
+	return ch, nil
+}
+
+// cancel sends an error to the subscriber.
+func (c *Channel) cancel(id uint32, err error) {
+	if ch, _ := c.removeResponseChannel(id); ch != nil {
+		ch <- Response{Error: err}
+	}
+}
+
+func (c *Channel) publish(id uint32, status *protostatus.Status, resp agentpb.AgentResponsePayload) {
+	if status != nil {
+		l := logger.Get(c.s.Context())
+		l.Errorf("got response %v with status %v", resp, status)
+		c.cancel(id, grpcstatus.FromProto(status).Err())
+		return
+	}
+	ch, err := c.removeResponseChannel(id)
+	if err != nil {
+		c.close(err)
+		return
+	}
+	if ch != nil {
```

Comment:
ch can't be nil here.

Reply:
if c.responses is closed, then it can be nil. https://github.com/percona/pmm-managed/pull/738/files/2ab2bafccc9d07ca3ecfb2f4a98567c47fb10966#diff-3bf1427f6ae504a050f61c903d63beb1ea7c927d6fa2a7b40035928728b25989R306